### PR TITLE
LibLine: Implement some history searching stuff

### PR DESCRIPTION
**Pull request by alimpfard**
This commit adds searching in the editor history with ^R.
It does so by instantiating...another Line::Editor inside the current
Line::Editor :^)

and yes, @linusg, this works just fine with js's syntax highlighting :^)

- pressing Enter while the search is up will execute the selected history element
- pressing Tab will quit search, but keep the buffer and continue the prompt (better key for this?)
- ^R and backspace will cycle backwards and forwards, respectively.

### CODE REVIEW

**Comment by linusg** [comment]
img

**Comment by bugaevc** [comment]
Can we also support [this feature](https://codeinthehole.com/tips/the-most-important-command-line-tip-incremental-history-searching-with-inputrc/) as an alternative way of searching through history? Without any config file editing to enable it, of course.

**Comment by alimpfard** [comment]
I had thought about that, but I skipped over it...
My thought process was "what if I *didn't* want to search for whatever that
was in the editor buffer?"

If you have any suggestions on how it should behave in that regard, let me
know and I'll add this search mode

**Comment by bugaevc** [comment]
> My thought process was "what if I *didn't* want to search for whatever that was in the editor buffer?"

Sorry, I don't follow: if you didn't want to search for it, you wouldn't have inputted it and pressed Up?

**Comment by alimpfard** [comment]
Imagine this scenario:

I am writing some random command when I remember that I have it already in
the history, but it won't match what I have written up to this point.
Either I erase to some point (ew)
Or I rewrite it up to that point (a waste)
Or...?

**Comment by alimpfard** [comment]
While writing that I realised it could just search for the substring that
the cursor marks the end of...

I think that's a good enough semantic for it, I'll implement it.

**Comment by bugaevc** [comment]
> I am writing some random command when I remember that I have it already in
the history, but it won't match what I have written up to this point.

You erase it, yes, using some clever key combo for "erase to beginning of the line". I personally just press Ctrl-C, which prints `^C` and puts me on a fresh line. And — is it any different than what you'd do with Ctrl-R?

**Comment by alimpfard** [comment]
> You erase it
Gosh, that's so inefficient. Well, the scheme I came up with should cover this as well, so I guess that's fine?

**Comment by bugaevc** [comment]
> Gosh, that's so inefficient

I still don't see how that's different from when you use Ctrl-R for searching?

> Well, the scheme I came up with should cover this as well, so I guess that's fine?

Maybe if you explain how your scheme is different from what libreadline does again :D

**Comment by alimpfard** [comment]
I legit have no idea what libreadline does!
I don't think readline even supports searches like that

**Comment by alimpfard** [comment]
> different from ^R

The comparison isn't with ^R, that's its own thing.
I'm just thinking about how we should handle normal "scrolling" in the history.

**Comment by bugaevc** [comment]
> I legit have no idea what libreadline does! I don't think readline even supports searches like that

Well, please do try it out. See the post I linked about how to enable it.

> I'm just thinking about how we should handle normal "scrolling" in the history

Basically if you start pressing up/down without typing anything, it, naturally, matches *every* line in history, so you get all your history entries.

File: Libraries/LibLine/Editor.cpp [hunk 1]
```diff
@@ -33,8 +33,9 @@
 
 namespace Line {
 
-Editor::Editor()
+Editor::Editor(bool lazy)
 {
+    m_is_lazy = lazy;
```
Comment (awesomekling) [inline Libraries/LibLine/Editor.cpp:38]: What does lazy mean in this context?

File: Libraries/LibLine/Editor.h [hunk 2]
```diff
@@ -197,6 +216,15 @@ class Editor {
     void recalculate_origin();
     void reposition_cursor();
 
+    bool m_finish = false;
+
+    OwnPtr<Editor> m_search_editor = nullptr;
```
Comment (awesomekling) [inline Libraries/LibLine/Editor.h:222]: No need to explicitly initialize smart pointers.

File: Libraries/LibLine/Editor.h [hunk 3]
```diff
@@ -197,6 +216,15 @@ class Editor {
     void recalculate_origin();
     void reposition_cursor();
 
+    bool m_finish = false;
```
Comment (awesomekling) [inline Libraries/LibLine/Editor.h:220]: Mixing `=` and `{}` initialization.

File: Libraries/LibLine/Editor.cpp (hunk 1 above)
Comment (alimpfard) [reply Libraries/LibLine/Editor.cpp:38]: Whether it will draw lazily (only whatever is needed)
I guess I'll rename it to always_refresh

**Comment by alimpfard** [comment]
hmmm...that went two commits back...

**Comment by bugaevc** [comment]
* In Ctrl-R mode, pressing Enter should not result in the command from history getting executed, you should be able to edit it furter and then press enter later to execute it as you do normally.
* When normally editing a command, pressing Ctrl-C should not result in it getting executed; we should just drop the command, without executing it or adding it to the history.

**Comment by alimpfard** [comment]
Ctrl-C executes? how did I mess that up?

But...what if I wanted it to?
I added pressing tab if further editing is desired, which makes a tiny bit of sense
if I go and swap those (enter to edit, tab to execute), it'll just be weird

**Comment by bugaevc** [comment]
> In Ctrl-R mode, pressing Enter should not result in the command from history getting executed, you should be able to edit it furter and then press enter later to execute it as you do normally.

> I added pressing tab if further editing is desired, which makes a tiny bit of sense

I take that back, in Bash + libreadline it does get executed, and Tab lets you edit it further.

**Comment by alimpfard** [comment]
arrrg @alimpfard

**Pull request merged**